pkg/aws/eni/node_manager.go

```diff
@@ -71,31 +73,46 @@ type NodeManager struct {
 	metricsAPI      metricsAPI
 	resyncTrigger   *trigger.Trigger
 	deficitResolver *trigger.Trigger
+	parallelWorkers int64
 }
 
 // NewNodeManager returns a new NodeManager
-func NewNodeManager(instancesAPI nodeManagerAPI, ec2API ec2API, k8sAPI k8sAPI, metrics metricsAPI) (*NodeManager, error) {
+func NewNodeManager(instancesAPI nodeManagerAPI, ec2API ec2API, k8sAPI k8sAPI, metrics metricsAPI, parallelWorkers int64) (*NodeManager, error) {
+	if parallelWorkers < 1 {
+		parallelWorkers = 1
+	}
+
 	mngr := &NodeManager{
-		nodes:        nodeMap{},
-		instancesAPI: instancesAPI,
-		ec2API:       ec2API,
-		k8sAPI:       k8sAPI,
-		metricsAPI:   metrics,
+		nodes:           nodeMap{},
+		instancesAPI:    instancesAPI,
+		ec2API:          ec2API,
+		k8sAPI:          k8sAPI,
+		metricsAPI:      metrics,
+		parallelWorkers: parallelWorkers,
 	}
 
 	deficitResolver, err := trigger.NewTrigger(trigger.Parameters{
 		Name:        "eni-node-manager-deficit-resolver",
 		MinInterval: time.Second,
 		TriggerFunc: func(reasons []string) {
+			sem := semaphore.NewWeighted(parallelWorkers)
+
 			for _, name := range reasons {
-				if node := mngr.Get(name); node != nil {
-					if err := node.ResolveIPDeficit(); err != nil {
-						node.logger().WithError(err).Warning("Unable to resolve IP deficit of node")
+				sem.Acquire(context.TODO(), 1)
```

Comment:
This can potentially block forever if other operations fail to release the semaphore since you are using `context.TODO()` - is it possible that this may get stuck? Perhaps have some long timeout associated with the context?

Reply:
Any lock can dead-lock if it isn't unlocked. Do you want to change how we deal with locks in general?

Reply:
Context provides a way of implementing some sort of deadline for potentially long-lived or cancelable operations. I was wondering whether it'd be useful to have some long timeout to signal that something is wrong and write an error, since we get that for free in this case, as the semaphore implementation in golang has us use context. 

I'd actually argue that anywhere we do something that may be bounded by some long-running operation (e.g., compilations, API interactions, etc.), that we should provide context so that we do not have a single goroutine holding a resource for a long amount of time when other goroutines may be contending on it.

But, it looks like in this case that there isn't a long running operation triggered by the subsequent code which releases the semaphore, so it isn't that big of a deal.